### PR TITLE
Transform

### PR DESCRIPTION
**Pull request by fraser-dunlop**
transform :: (a -> a) -> b -> b

Should work for all conjure types b. Currently (a -> a) is only allowed to be a function type.

### CODE REVIEW

File: src/Conjure/Language/Expression/Op/Transform.hs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+instance ToJSON    x => ToJSON    (OpTransform x) where toJSON = genericToJSON jsonOptions
+instance FromJSON  x => FromJSON  (OpTransform x) where parseJSON = genericParseJSON jsonOptions
+
+instance (TypeOf x, Pretty x) => TypeOf (OpTransform x) where
+    typeOf p@(OpTransform f x) = do
+        tyF <- typeOf f
+        (from, to) <- case tyF of
+            TypeFunction from to   -> return (from, to)
+            TypeSequence      to   -> return (TypeInt TagInt, to)
+            TypeRelation [from,to] -> return (from, to)
```
Comment (ozgurakgun) [inline src/Conjure/Language/Expression/Op/Transform.hs]: I don't think relations make sense here.

File: src/Conjure/Language/Type.hs [hunk 2]
```diff
@@ -239,3 +243,47 @@ typeCanIndexMatrix TypeBool{} = True
 typeCanIndexMatrix TypeInt {} = True
 typeCanIndexMatrix TypeEnum{} = True
 typeCanIndexMatrix _          = False
+
+morphing :: (?typeCheckerMode :: TypeCheckerMode)
```
Comment (ozgurakgun) [inline src/Conjure/Language/Type.hs]: Can you write a one line description of this function please? Not sure I understand what it achieves.

File: src/Conjure/Language/Expression/Op/Transform.hs (hunk 1 above)
Comment (ozgurakgun) [inline src/Conjure/Language/Expression/Op/Transform.hs]: I don't think relations make sense here.

**Comment by ozgurakgun** [comment]
Moreover: the tests don't seem right. You removed the expected output of several tests?

File: src/Conjure/Language/Type.hs (hunk 2 above)
Comment (fraser-dunlop) [reply src/Conjure/Language/Type.hs]: It is just a helper for extracting the type that a homomorphism is "morphing". Right now it is mostly a placeholder - there will be cases for Permutation and Sequence once they are supported.

File: src/Conjure/Language/Expression/Op/Transform.hs (hunk 1 above)
Comment (fraser-dunlop) [reply src/Conjure/Language/Expression/Op/Transform.hs]: Agreed

**Comment by fraser-dunlop** [comment]
I accidentally nuked the tests - how embarrassing. Unless you have a better suggestion I will just copy over tests from master since transform *shouldn't* affect the results of any of those.

**Comment by ozgurakgun** [comment]
That is fine as long as you run the LIMIT_TIME=0 version of the whole test suite to make sure nothing is broken?

**Comment by ozgurakgun** [comment]
Or `--limit-time=0` depending on how you are running the tests.

**Comment by fraser-dunlop** [comment]
> That is fine as long as you run the LIMIT_TIME=0 version of the whole test suite to make sure nothing is broken?

Happy to report that nothing is broken.

**Comment by ozgurakgun** [comment]
Great news, I'll look at the diff.

**Comment by ozgurakgun** [comment]
Some of those changes make me think you branched from an earlier version of master. I'll merge master on top of this, let's see if we get a smaller diff.

**Comment by ozgurakgun** [comment]
Looks great. Do you want to do map/reduce/filter etc next or the lambda stuff? :)

Regardless, do you want to merge this or keep working on this branch and we can merge all of them in one go? I guess this is the most important for the permutations branch, so I am happy to merge this now if you like. We can do/merge the rest later.

**Comment by fraser-dunlop** [comment]
I'd like to merge this now since it works on its own. I think pull requests for each feature is a nice workflow rather than all in one go.

My TODO list is something like [permutations, lex-ordering, int-strengthening, map, filter, fold, lambdas] which can all be done reasonably independently.

**Comment by ozgurakgun** [comment]
I know all of that - except not sure about lex-ordering. Do you mean the optimisation thing? That's easy, even I can do that :)

**Comment by fraser-dunlop** [comment]
Merged! Yay!

> I know all of that - except not sure about lex-ordering. Do you mean the optimisation thing? That's easy, even I can do that :)

I meant lex-flattenning - there is an implementation on the permutation-minimal branch but I have my suspicions about it - worth doing a PR just for that imo.

**Pull request merged**